### PR TITLE
Feat/multi network relayer

### LINKED ISSUE #402: Replace Web3 by Ethers in relayer-cli, use TransactionBatcher directly
[TransactionBatcher.sol](https://github.com/kleros/kleros-v2/blob/e67fc2a8343d79d5f63256d8246a80f1f7366d2b/contracts/src/utils/TransactionBatcher.sol#L6)

### PR DESCRIPTION
**Pull request by mani99brar**
#### Agnostic relayer-cli:
<ul> Merged testnetRelayer and devnetRelayer into relayer which runs for both networks and multiple chains, with relayer config for devnetSenders/testnetSenders env var:

1) no value passed represents respective network will not run
2) first address passed = zeroAddress represents relay for all addresses on the respective network
3) address/addresses represents relay for the passed addresses only
             

</ul>

#### Added TransactionBatcher.sol [from kleros-v2](https://github.com/kleros/kleros-v2/blob/a26fe616ffe284841ec99df8eebd4a29e4c074a2/contracts/src/utils/TransactionBatcher.sol#L4):
<ul>Replaced deprecated web3-batcher with TransactionBatcher.sol  </ul>

#### Added validation and support for pm2
<ul>Added type and env checks to ensure strict typescript along with more custom errors</ul>

## Summary by CodeRabbit

- **New Features**
	- Launched a new smart contract enabling batch transactions with options for strict or lenient execution.
	- Introduced an enhanced relayer service with multi-network support and detailed event logging.
	- Added custom error handling classes for improved error reporting.
	- Implemented new interfaces and utilities for better message handling and network configuration.
	- Enhanced logging capabilities for message relaying operations and error handling.

- **Refactor**
	- Streamlined configuration settings for flexible management of sender addresses, chain IDs, and network parameters.
	- Improved error handling and connection logic for more robust transaction and message relaying.
	- Enhanced module resolution and organization for better code clarity.
	- Updated function signatures and logic for improved modularity and error handling.

- **Chores**
	- Updated scripts, dependencies, and build configurations to boost system performance.

- **Tests**
	- Refined testing workflows to verify batched message processing and error management.
	- Expanded test coverage for process termination signals and file cleanup operations.

---

## PR-Codex overview
This PR focuses on refactoring and enhancing the `relayer-cli` codebase, including changes to configuration, error handling, and message relaying functionality. It introduces new features and improves existing ones while updating dependencies.

### Detailed summary
- Removed `testnetRelayer.ts` and `devnetRelayExample.ts`.
- Updated `jest.config.ts` to include `moduleNameMapper`.
- Changed `baseUrl` in `tsconfig.json` to `.` and added `outDir`.
- Renamed `devnet-relayer` to `vea-relayer` in `ecosystem.config.js`.
- Added new `BotEvents` for message relay in `botEvents.ts`.
- Enhanced logging in `logger.ts` for message relay events.
- Introduced new error classes in `errors.ts`.
- Added `TransactionBatcher` contract in `TransactionBatcher.sol`.
- Updated `.env.dist` to reflect new sender address format and removed old entries.
- Modified `proof.ts` to include new interfaces for message data.
- Refactored functions in `relayer.ts` for clearer structure and added new relaying logic.
- Improved `relayerHelpers.ts` with new functions for cleanup and configuration management.
- Updated `relay.ts` to support batching and relaying messages with improved error handling.
- Enhanced tests in `relay.test.ts` to cover new functionalities and edge cases.
- Updated `bridgeRoutes.ts` to use a new structure for contract addresses and added error handling.
- Updated `ethers.ts` to support new contract factories and improved address handling.
- General cleanup and organization of code for better maintainability.

> The following files were skipped due to too many changes: `yarn.lock`

> ✨ Ask PR-Codex anything about this PR by commenting with `/codex {your question}`

### CODE REVIEW

File: relayer-cli/.env.dist [hunk 1]
```diff
@@ -1,22 +1,20 @@
...
-VEAINBOX_ARBSEPOLIA_TO_SEPOLIA_ADDRESS=0x906dE43dBef27639b1688Ac46532a16dc07Ce410
-VEAINBOX_ARBSEPOLIA_TO_CHIADO_ADDRESS=0xAb53e341121448Ae259Da8fa17f216Cb0e21199C
-VEAOUTBOX_ARBSEPOLIA_TO_SEPOLIA_ADDRESS=0x906dE43dBef27639b1688Ac46532a16dc07Ce410
-VEAOUTBOX_ARBSEPOLIA_TO_CHIADO_ADDRESS=0xAb53e341121448Ae259Da8fa17f216Cb0e21199C
-
-# Subgraph endpoints, Example: "85918/vea-inbox-arb-sepolia-devnet/version/latest"
-VEAINBOX_ARBSEPOLIA_TO_SEPOLIA_SUBGRAPH=11111/your-subgraph/version/your-version
-VEAINBOX_ARBSEPOLIA_TO_CHIADO_SUBGRAPH=11111/your-subgraph/version/your-version
+# Subgraph endpoint, Example: "85918/vea-inbox-arb-sepolia-devnet/version/latest"
+RELAYER_SUBRAPGH="61738/vea-inbox-arb-sepolia/version/latest"
```
[the remark this answers is not in the document]
Comment (mani99brar) [reply relayer-cli/.env.dist]: Done

File: contracts/src/utils/TransactionBatcher.sol [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+    function batchSendUnchecked(
+        address[] memory targets,
+        uint256[] memory values,
+        bytes[] memory datas
+    ) public payable {
+        for (uint256 i = 0; i < targets.length; i++) {
+            targets[i].call{value: values[i]}(datas[i]); // Intentionally ignoring return value.
+        }
+    }
```
[the remark this answers is not in the document]
Comment (mani99brar) [reply contracts/src/utils/TransactionBatcher.sol:22]: Out of scope

File: contracts/src/utils/TransactionBatcher.sol [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+// Adapted from https://github.com/daostack/web3-transaction-batcher/blob/1b88d2ea062f8f2d9fdfdf9bbe85d2bbef780151/contracts/Batcher.sol
+
+contract TransactionBatcher {
+    function batchSend(address[] memory targets, uint256[] memory values, bytes[] memory datas) public payable {
+        for (uint256 i = 0; i < targets.length; i++) {
+            (bool success, ) = targets[i].call{value: values[i]}(datas[i]);
+            if (!success) revert("transaction failed"); // All the calls must succeed.
+        }
+    }
```
[the remark this answers is not in the document]
Comment (mani99brar) [reply contracts/src/utils/TransactionBatcher.sol:12]: Out of scope

File: relayer-cli/src/utils/proof.ts [hunk 4]
```diff
@@ -27,14 +37,22 @@ const getMessageDataToRelay = async (
                 data
                 }
             }`
-    );
+    )) as MessageSentsDataResponse;
 
     return [result[`messageSents`][0].to.id, result[`messageSents`][0].data];
```
[the remark this answers is not in the document]
Comment (mani99brar) [reply relayer-cli/src/utils/proof.ts:42]: Let it fail if now result is found

File: relayer-cli/src/consts/bridgeRoutes.ts [hunk 5]
```diff
@@ -1,61 +1,88 @@
...
-    veaInboxAddress: process.env.VEAINBOX_ARBSEPOLIA_TO_CHIADO_ADDRESS,
-    veaOutboxAddress: process.env.VEAOUTBOX_ARBSEPOLIA_TO_CHIADO_ADDRESS,
-    batcher: process.env.TRANSACTION_BATCHER_CONTRACT_ADDRESS_CHIADO,
-    rpcOutbox: process.env.RPC_CHIADO,
-    veaOutboxContract: veaOutboxArbToGnosisContract,
+    veaContracts: arbToGnosisContracts,
+    batcherAddress: process.env.TRANSACTION_BATCHER_CONTRACT_CHIADO!,
+    rpcOutbox: process.env.RPC_CHIADO!,
   },
 };
```
[the remark this answers is not in the document]
Comment (mani99brar) [reply relayer-cli/src/consts/bridgeRoutes.ts:73]: General env validation will be added with dockerization

File: relayer-cli/src/utils/errors.ts
Comment (fcanela) [inline relayer-cli/src/utils/errors.ts:1]: This error classes a great!

As a tiny tiny tiny detail, I would try to normalize the class name: there is one that finishes with `Error` while the rest doesn't. In Atlas, we always try to put the suffix on error classes. You can also go the other way and change `DataError` for `InvalidCallData`.

File: relayer-cli/src/relayer.ts
Comment (fcanela) [inline relayer-cli/src/relayer.ts:1]: You did a great job documenting this file and using the EventEmitter pattern to split concerns!

File: relayer-cli/src/relayer.ts [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+ * @param config.networkConfigs The network configurations retrieved from the env.
+ * @param config.shutdownManager The shutdown manager
+ * @param config.emitter The event emitter
+ */
+export async function start({ networkConfigs, shutdownManager, emitter }: RelayerConfig) {
+  initializeEmitter(emitter);
+  let delayAmount = 7200 * 1000; // 2 hours in ms
+  while (!shutdownManager.getIsShuttingDown()) {
+    for (const networkConfig of networkConfigs) {
+      delayAmount = await processNetworkConfig(networkConfig, shutdownManager, emitter, delayAmount);
```
Comment (fcanela) [inline relayer-cli/src/relayer.ts:35]: It's fine, but it is surprising that `processNetworkConfig` is returning a delayAmount. I would have not expected that from the function name. No need to change anything though

File: relayer-cli/src/relayer.ts [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+    ? await relayBatch({ chainId, network, nonce, maxBatchSize, emitter })
+    : await relayAllFrom(chainId, network, nonce, senders, emitter);
+
+  if (nonce == null) return currentDelay;
+
+  await updateStateFile(chainId, Math.floor(Date.now() / 1000), nonce, network, emitter);
+
+  if (network === Network.DEVNET) {
+    return 1000 * 10; // 10 seconds for devnet
+  } else {
```
Comment (fcanela) [inline relayer-cli/src/relayer.ts:76]: This else branch is not needed. The `if` already has a return and can act as a guard.

File: relayer-cli/src/utils/ethers.ts [hunk 8]
```diff
@@ -35,58 +35,83 @@ function getVeaInboxProvider(veaInboxAddress: string, privateKey: string, rpc: J
...
-    case "gnosis":
-      return VeaOutboxArbToGnosis__factory.connect(veaOutboxAddress, getWallet(privateKey, web3ProviderURL));
+    case "sepolia": {
+      switch (network) {
+        case Network.DEVNET:
+          return VeaOutboxArbToEthDevnet__factory.connect(veaOutboxAddress, getWallet(privateKey, web3ProviderURL));
+        case Network.TESTNET:
+          return VeaOutboxArbToEth__factory.connect(veaOutboxAddress, getWallet(privateKey, web3ProviderURL));
+        default:
+          throw new Error(`Unsupported network for sepolia: ${network}`);
```
Comment (fcanela) [inline relayer-cli/src/utils/ethers.ts:62]: This can also be extracted in a `UnsupportedNetworkError` (reusable for both sepolia and chiado) to make the code easier to test

File: relayer-cli/src/utils/ethers.ts [hunk 9]
```diff
@@ -35,58 +35,83 @@ function getVeaInboxProvider(veaInboxAddress: string, privateKey: string, rpc: J
...
-    case "gnosis":
-      return VeaOutboxArbToGnosis__factory.connect(veaOutboxAddress, getWalletRPC(privateKey, rpc));
+    case "sepolia": {
+      switch (network) {
+        case Network.DEVNET:
+          return VeaOutboxArbToEthDevnet__factory.connect(veaOutboxAddress, getWalletRPC(privateKey, rpc));
+        case Network.TESTNET:
+          return VeaOutboxArbToEth__factory.connect(veaOutboxAddress, getWalletRPC(privateKey, rpc));
+        default:
+          throw new Error(`Unsupported network for sepolia: ${network}`);
```
Comment (fcanela) [inline relayer-cli/src/utils/ethers.ts:96]: (Same)

File: relayer-cli/src/utils/relay.test.ts
Comment (fcanela) [inline relayer-cli/src/utils/relay.test.ts:1]: This test looks good, but the amount of mocking invites to think about using integration tests (using hardhat and real data, if available) instead of unit. I assume that there is no real scenarios yet to have a realistic test?

No need to change anything, I just want to understand the context and see if we can make robust tests with the less effort required.

File: relayer-cli/src/utils/relay.ts [hunk 10]
```diff
@@ -35,101 +39,113 @@ const getCount = async (veaOutbox: VeaOutboxArbToEth | VeaOutboxArbToGnosis, cha
...
+        const callData = veaOutbox.interface.encodeFunctionData("sendMessage", [proof, nonce, to, data]);
+        datas.push(callData);
+        targets.push(veaOutboxAddress);
+        values.push(0);
+        batchMessages += 1;
+        nonce++;
       } catch {
+        emitter.emit(BotEvents.MESSAGE_EXECUTION_FAILED, nonce);
         nonce++;
-        continue;
```
Comment (fcanela) [inline relayer-cli/src/utils/relay.ts:121]: I am not sure what happens when a `MESSAGE_EXECUTION_FAILED` event is emitted, but it will be for sure a better alternative that just swallowing the errors in a silent `continue`. Nice refactor here.

Any reason not to capture the error and provide that context to the emitted event? We may benefit from seeing a stack trace and/or error message to figure out what went wrong.

File: relayer-cli/src/utils/relay.ts [hunk 11]
```diff
@@ -142,45 +158,69 @@ const relayBatch = async ({
...
+const relayAllFrom = async (
+  chainId: number,
+  network: Network,
+  nonce: number,
+  msgSenders: string[],
+  emitter: EventEmitter
+): Promise<number | null> => {
+  const bridgeConfig = getBridgeConfig(chainId);
+  const { veaContracts, batcherAddress, rpcOutbox } = bridgeConfig;
+  const privateKey = process.env.PRIVATE_KEY;
```
Comment (fcanela) [inline relayer-cli/src/utils/relay.ts:170]: This is not bad, but the code is usually easier to test if instead of reaching out of the function, you allow this configuration parameter to be injected (with a fallback for convenience, if you want).

```typescript

const relayAllFrom = async (
  chainId: number,
  // ...
  privateKey: string = process.env.PRIVATE_KEY,
): Promise<number | null> => {
```

You can use this dependency/configuration injection pattern also with all the methods that the function is reaching for and you may want to mock during testing.

**Review by fcanela — COMMENTED** [review]
Great job! Code looks nice. I've added a few notes here and there, but they are just suggestions and mostly opinions on code practices.

I'm not approving it as I'm not familiar enough with Vea to provide such approval, but the refactor looks good to me!

File: relayer-cli/src/relayer.ts (hunk 6 above)
Comment (mani99brar) [reply relayer-cli/src/relayer.ts:35]: Will consider this for future when naming functions, added the return in JSDoc comment.

File: relayer-cli/src/utils/errors.ts
Comment (mani99brar) [reply relayer-cli/src/utils/errors.ts:1]: Will use InvalidCallData 👍

File: relayer-cli/src/utils/ethers.ts (hunk 8 above)
Comment (mani99brar) [reply relayer-cli/src/utils/ethers.ts:62]: I agree 👍 , missed to update this error

File: relayer-cli/src/utils/relay.test.ts
Comment (mani99brar) [reply relayer-cli/src/utils/relay.test.ts:1]: For real data I have sent multiple messages across the bridge we can fork and run tests for it. Also with local contract and subgraph deployments we can achieve integration testing without adding any contract mocks as the relayer is independent of unhappy or happy path, the unhappy path requires mocks for canonical bridging.

File: relayer-cli/src/utils/relay.ts (hunk 10 above)
Comment (mani99brar) [reply relayer-cli/src/utils/relay.ts:121]: Ideally we dont want a batch to fail for a single failing msg execution which depends on the destination contract of the message. I added this event in case we want to monitor failing messages and add a retry mechanism for them after some time. I dont think the error message would be relevant here as it would always be a result of failed execution on the destination contract OR when the bot runs out of gas.

File: relayer-cli/src/utils/relay.ts (hunk 11 above)
Comment (mani99brar) [reply relayer-cli/src/utils/relay.ts:170]: Got it 👍

**Review by jaybuidl — DISMISSED** [review]

**Review by jaybuidl — APPROVED** [review]

**Comment by jaybuidl** [comment]
The veascan build issues are addressed in https://github.com/kleros/vea/pull/418
